src/sentry/integrations/bitbucket/client.py

```diff
@@ -0,0 +1,203 @@
+from __future__ import absolute_import
+
+import datetime
+import jwt
+
+from unidiff import PatchSet
+
+from six.moves.urllib.parse import urlparse
+
+from sentry.utils.http import absolute_uri
+from sentry.integrations.atlassian_connect import get_query_hash
+from sentry.integrations.client import ApiClient
+
+BITBUCKET_KEY = '%s.bitbucket' % urlparse(absolute_uri()).hostname
+
+
+class BitbucketAPIPath(object):
+    """
+    All UUID's must be surrounded by culybraces.
+    repo_slug - repository slug or UUID
+    username - username or UUID
+    """
+
+    issue = u'/2.0/repositories/{username}/{repo_slug}/issues/{issue_id}'
+    issues = u'/2.0/repositories/{username}/{repo_slug}/issues'
+    issue_comments = u'/2.0/repositories/{username}/{repo_slug}/issues/{issue_id}/comments'
+
+    repository = u'/2.0/repositories/{username}/{repo_slug}'
+    repository_commits = u'/2.0/repositories/{username}/{repo_slug}/commits/{revision}'
+    repository_diff = u'/2.0/repositories/{username}/{repo_slug}/diff/{spec}'
+    repository_hook = u'/2.0/repositories/{username}/{repo_slug}/hooks/{uid}'
+    repository_hooks = u'/2.0/repositories/{username}/{repo_slug}/hooks'
+
+
+class BitbucketAPI(ApiClient):
+
+    def __init__(self, base_url, shared_secret, subject):
+        # subject is probably the clientKey
+        super(BitbucketAPI, self).__init__(verify_ssl=False)
+        self.base_url = base_url
+        self.shared_secret = shared_secret
+        self.subject = subject
+
+    # TODO(LB): do has_auth and bind_auth belong here?
+
+    def request(self, method, path, data=None, params=None, **kwargs):
+        jwt_payload = {
+            'iss': BITBUCKET_KEY,
+            'iat': datetime.datetime.utcnow(),
+            'exp': datetime.datetime.utcnow() + datetime.timedelta(seconds=5 * 60),
+            'qsh': get_query_hash(path, method.upper(), params),
+            'sub': self.subject,
+        }
+        encoded_jwt = jwt.encode(jwt_payload, self.shared_secret)
+        params = dict(**(params or {}))
```

Comment:
Might be my lack of Python-fu, but is this different than just setting the argument to `data={}` in the method signature?

Reply:
So, I really hesitate to put dynamic structures in the method signature, due to things being passed by reference, and there are tons of possibilities for bugs. BUT that totally used to do something; there used to be something I was adding to params, and now I'm not. I'll just change this to passing whatever is in params now! Either way good catch!!!

Reply:
@mnoble 
More to your question pretty sure the pythonic way to handle something like that is:
```
def foo(data=None):
   if data is None:
       data = {}
```
putting dynamic structures in the method signature tends to result in common gotchas! 
http://docs.python-guide.org/en/latest/writing/gotchas/

Reply:
Makes sense. I don't think it's really a concern when you're using literals as the default value, though.

Reply:
![](https://i.imgur.com/WrKPhfd.gif)